### PR TITLE
[SPARK-7989][Core][Tests] Fix flaky tests in ExternalShuffleServiceSuite and SparkListenerWithClusterSuite

### PR DESCRIPTION
**Pull request by zsxwing**
The flaky tests in ExternalShuffleServiceSuite and SparkListenerWithClusterSuite will fail if there are not enough executors up before running the jobs.

This PR adds `JobProgressListener.waitUntilExecutorsUp`. The tests for the cluster mode can use it to wait until the expected executors are up.

### CODE REVIEW

**Comment by srowen** [comment]
@zsxwing are these things you identify actually causing test failures for you or in Jenkins or is this theoretical?

**Comment by zsxwing** [comment]
@srowen These two failures happened in my PR #6457. I fixed them in #6457 to pass Jenkins. But I think they are not a part of #6457 and it's better to fix them in a separate PR.

**Comment by zsxwing** [comment]
retest this please

**Comment by squito** [comment]
if this is just for testing, how about we put it into a SparkListener in the tests instead, and avoid changing `JobProgressListener`?  You could put it into a common listener trait that could be used for both tests.  The listener would just count how many times `onExecutorAdded` / `onExecutorRemoved` were called and would have a similar `waitUntilExecutorsUp` method.

**Comment by zsxwing** [comment]
> if this is just for testing, how about we put it into a SparkListener in the tests instead, and avoid changing JobProgressListener? You could put it into a common listener trait that could be used for both tests. The listener would just count how many times onExecutorAdded / onExecutorRemoved were called and would have a similar waitUntilExecutorsUp method.

The problem of a new SparkListener is "adding an Executor" may happen before "adding the new SparkListener to SparkContext". So the new SparkListener may miss some messages before adding itself to SparkContext. `JobProgressListener` is added as early as possible so that it won't miss any message.

**Comment by squito** [comment]
> The problem of a new SparkListener is "adding an Executor" may happen before "adding the new 
> SparkListener to SparkContext". So the new SparkListener may miss some messages before adding 
> itself to SparkContext. JobProgressListener is added as early as possible so that it won't miss any 
> message.

hmm, good point.  You can make sure listeners are added immediately with [`spark.extraListeners`](https://github.com/apache/spark/blob/90c606925e7ec8f65f28e2290a0048f64af8c6a6/core/src/main/scala/org/apache/spark/SparkContext.scala#L1945), but that is admittedly a pain since you don't get a handle on our listener, so you need some to communicate via some global variable.  I guess I don't feel strongly either way then.

**Comment by andrewor14** [comment]
retest this please

**Comment by andrewor14** [comment]
@zsxwing why does this reduce the flakiness? What is the source of the flakiness?

File: core/src/test/scala/org/apache/spark/ExternalShuffleServiceSuite.scala [hunk 1]
```diff
@@ -55,6 +55,14 @@ class ExternalShuffleServiceSuite extends ShuffleSuite with BeforeAndAfterAll {
     sc.env.blockManager.externalShuffleServiceEnabled should equal(true)
     sc.env.blockManager.shuffleClient.getClass should equal(classOf[ExternalShuffleClient])
 
+    // In a slow machine, one slave may register hundreds of milliseconds ahead of the other one.
+    // If we don't wait for all salves, it's possible that only one executor runs all jobs. Then
```
Comment (andrewor14) [inline core/src/test/scala/org/apache/spark/ExternalShuffleServiceSuite.scala]: salves

File: core/src/test/scala/org/apache/spark/ExternalShuffleServiceSuite.scala [hunk 2]
```diff
@@ -55,6 +55,14 @@ class ExternalShuffleServiceSuite extends ShuffleSuite with BeforeAndAfterAll {
...
+    // all shuffle blocks will be in this executor, ShuffleBlockFetcherIterator will directly fetch
+    // local blocks from the local BlockManager and won't send requests to ExternalShuffleService.
+    // In this case, we won't receive FetchFailed. And it will make this test fail.
+    // Therefore, we should wait until all salves are up
```
Comment (andrewor14) [inline core/src/test/scala/org/apache/spark/ExternalShuffleServiceSuite.scala]: salves again

**Comment by andrewor14** [comment]
@zsxwing IIUC this only adds some wait logic for the two test suites, but the intent is that we don't change the logic in `BroadcastSuite` right?

**Comment by andrewor14** [comment]
master 1.4

**Comment by zsxwing** [comment]
> @zsxwing IIUC this only adds some wait logic for the two test suites, but the intent is that we don't change the logic in BroadcastSuite right?

Right. `BroadcastSuite` could also use this new method, so I updated it to simplify codes. And thanks for fixing the typos.

**Pull request closed without merging**